### PR TITLE
Python: use separate output for tkinter

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

See issue #19255 and #1819.
- Unify expressions of different interpreters
- Reduce closure size of Python 3.x interpreters
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10324,11 +10323,12 @@ in
     inherit clangStdenv fetchurl fetchgit fetchpatch stdenv pkgconfig intltool freetype fontconfig
       libxslt expat libpng zlib perl mesa_drivers spice_protocol libunwind
       dbus libuuid openssl gperf m4 libevdev tradcpp libinput mcpp makeWrapper autoreconfHook
-      autoconf automake libtool xmlto asciidoc flex bison python mtdev pixman
+      autoconf automake libtool xmlto asciidoc flex bison mtdev pixman
       cairo epoxy;
     inherit (darwin) apple_sdk cf-private libobjc;
     bootstrap_cmds = if stdenv.isDarwin then darwin.bootstrap_cmds else null;
     mesa = mesa_noglu;
+    python = python27.override {x11Support = false;};
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: python = pythonWithoutX11;

**Comment by FRidh** [comment]
Sigh... putting curses in a separate output won't help because python and libpython directly link against it.

**Comment by domenkozar** [comment]
@FRidh that's fine I think. It's not ideal, but this way Python will be way less painful on Nix.

**Comment by RonnyPfannschmidt** [comment]
this adds a bonus, since pypy is not shipping tkinter but there is a pypi package for it based on cffi
(so nix could simply match that behaviour now)

**Comment by FRidh** [comment]
@domenkozar as I wrote to @copumpkin in https://github.com/NixOS/nixpkgs/issues/19225#issuecomment-252701145 I think this is a good balance between closure size and usability.

**Comment by domenkozar** [comment]
Nice. Very quick testing says this is very promising.

Fixes https://github.com/NixOS/nixpkgs/issues/492 and https://github.com/NixOS/nixpkgs/issues/7134

```
$ /nix/store/b69qg3a33p8pdlhwr4klwgq081wgdq26-python2.7-virtualenv-15.0.3/bin/virtualenv  fooooo
New python executable in /home/ielectric/dev/nixpkgs/fooooo/bin/python2.7
Also creating executable in /home/ielectric/dev/nixpkgs/fooooo/bin/python
Installing setuptools, pip, wheel...done.

$ cd fooooo/

$ bin/python
Python 2.7.12 (default, Jun 25 2016, 21:49:32) 
[GCC 5.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3
<module 'sqlite3' from '/nix/store/3vg24ijz3bc3fjsgpr8zarrsnfadg0f0-python-2.7.12/lib/python2.7/sqlite3/__init__.pyc'>
```

**Comment by FlorentBecker** [comment]
This breaks the scripts in scribus again (see #19292).

**Comment by FRidh** [comment]
> This breaks the scripts in scribus again (see #19292).

Correct, as long as we split Python we will break some things, but with `tkinter` its only very few packages that are affected. The fix will be `pythonEnv = python.withPackages(ps:[ps.python.tkinter])`.

**Comment by copumpkin** [comment]
@FRidh 👍! I'm wondering if we can alias `python.tkinter` to `pythonPackages.tkinter` so people don't have to know that it's bundled with python (and we have the option to move things around more).

**Comment by FRidh** [comment]
yes, we can simply add an alias. However, `tkinter` won't work exactly the same as Python packages. E.g., all Python packages have `python` as `propagatedBuildInput`, but `tkinter` doesn't. A simple fix would be to use `mkPythonDerivation` and then copy the contents of `python.tkinter`. But then `python.tkinter` and `pythonPackages.tkinter` would be slightly different.

**Comment by copumpkin** [comment]
@FRidh ah, not really sure there. My main issue is that many people probably don't know that tkinter is meant to be bundled with python, and my general approach when I find a failing dependency is to just look for it in pythonPackages. One possibility might be just to put this in pythonPackages:

```
{
  tkinter = error "tkinter isn't quite like other python derivations, so you should refer to it by python.tkinter instead of pythonPackages.tkinter";
}
```

That said, it would be nicest if it could look and feel like a conventional derivation.

**Comment by FRidh** [comment]
@copumpkin I added a proper Python package in `python-packages.nix`, see 331cd94b77f2551c01c634e8c20a880c62d56e97.

File: doc/languages-frameworks/python.md [hunk 2]
```diff
@@ -422,7 +422,7 @@ exists.
 
 #### Missing `tkinter` module standard library
 
-To reduce closure size the `Tkinter`/`tkinter` module is available as `python.tkinter`.
+To reduce closure size the `Tkinter`/`tkinter` is put in a separate output. The `tkinter` is available as `python35Packages.tkinter`.
```
Comment (copumpkin) [inline doc/languages-frameworks/python.md]: Perhaps just use `pythonPackages` here?

Comment (FRidh) [inline doc/languages-frameworks/python.md]: In these docs I mix a lot between the different sets. The idea behind it is that the reader sees that these things are valid for all interpreter versions. However, maybe it is better to just always stick with `pythonPackages` and mention only once that other sets can be used as well.

**Comment by vcunat** [comment]
So all those remaining `sqlite3` mentions should also get removed? This blocks staging evaluation ATM.

**Comment by FRidh** [comment]
@vcunat yes they should. I thought I got them all.

**Comment by FRidh** [comment]
ahh yes, it wasn't just pythonPackages.modules.sqlite3 but also pythonPackages.sqlite3. I forgot those.

**Comment by FRidh** [comment]
I'm sure I've asked this before but how can one test whether the whole of Nixpkgs evaluates?

**Comment by vcunat** [comment]
There are also `python27Full` and `curses`, maybe more.

**Comment by vcunat** [comment]
@FRidh: for quick eval I regularly use
```
nix-env -f . -qaP --meta --json --drv-path --show-trace >/dev/null
```

But one can do `nix-build pkgs/top-level/release.nix -A tarball -Q` to be proper (all Hydra platforms etc.).

**Comment by FRidh** [comment]
Thanks @vcunat . That should be it.

**Comment by vcunat** [comment]
No, there were quite a few more: 6eeea6e.

**Comment by FRidh** [comment]
hmm, it seemed to evaluate. But thank you for fixing the (hopefully) last ones.

**Comment by domenkozar** [comment]
Thank you @FRidh for doing this <3

**Comment by copumpkin** [comment]
Argh, a complication: the darwin stdenv bootstrap depends on python (because LLVM's build process uses it) and this change massively slowed down our bootstrap process, since it now needs to build a ton of TK stuff 😦 

Can we make the TK stuff optional so I can ask for a TK-free python during bootstrap? Then only the final "userspace" one would get TK.

**Comment by copumpkin** [comment]
Also, it actually seems to have broken the python build, which prevents bootstrap 😢

**Comment by vcunat** [comment]
@copumpkin: I think that should be solved by using `pythonSmall` for bootstrapping, as that one is without X, tcl and tk.

**Comment by FRidh** [comment]
I am actually thinking of getting rid of the tkinter output again, and just
build without tkinter by default, and then build once with tkinter and copy
the tkinter module in another derivation. Then we can get rid of
pythonSmall.

**Comment by FRidh** [comment]
@vcunat @copumpkin see https://github.com/NixOS/nixpkgs/pull/19594 , I think it's a better solution than what I merged earlier.

**Comment by copumpkin** [comment]
Dammit, I was just reaching for `pythonSmall` during bootstrap 😄 I don't really have a strong opinion on how Python actually works but I'll take a look at the other PR and opine if I have something useful to say.

**Pull request merged**